File: impl/cpp/tests/test_correctness.cpp
/**
 * test_correctness.cpp — correctness tests for the UOV C++ implementation.
 *
 * Build with CMake and run via ctest.  Each TEST_* function asserts its
 * invariants; any failure causes a non-zero exit code detected by ctest.
 */

#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include "uov/keygen.hpp"

static int passed = 0, failed = 0;

#define RUN_TEST(name) \
    do { \
        std::cout << "  " << #name << " ... "; std::cout.flush(); \
        try { name(); std::cout << "PASS\n"; ++passed; } \
        catch (std::exception& e) { std::cout << "FAIL: " << e.what() << "\n"; ++failed; } \
        catch (...) { std::cout << "FAIL (unknown)\n"; ++failed; } \
    } while(0)

static void CHECK(bool cond, const char* msg) {
    if (!cond) throw std::runtime_error(msg);
}

// ── eval_affine ───────────────────────────────────────────────────────────────

void test_eval_affine_exhaustive() {
    // Exhaustive over GF(7)^2 × GF(7)^2
    const int q = 7, o = 2, v = 2;
    std::mt19937 rng(0);
    auto comp = uov::CentralMapComp::random(q, o, v, rng);
    for (int o0 = 0; o0 < q; ++o0)
    for (int o1 = 0; o1 < q; ++o1)
    for (int v0 = 0; v0 < q; ++v0)
    for (int v1 = 0; v1 < q; ++v1) {
        uov::Vec oil{o0, o1}, vin{v0, v1};
        int lhs = comp.eval(oil, vin);
        auto lc = comp.lin_coeff(vin);
        int rhs = (uov::dot(oil, lc, q) + comp.vin_const(vin)) % q;
        CHECK(lhs == rhs, "eval_affine failed");
    }
}

void test_eval_affine_random() {
    const int q = 31, o = 4, v = 8;
    std::mt19937 rng(7);
    auto comp = uov::CentralMapComp::random(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    for (int t = 0; t < 200; ++t) {
        uov::Vec oil(o), vin(v);
        for (auto& x : oil) x = dist(rng);
        for (auto& x : vin) x = dist(rng);
        int lhs = comp.eval(oil, vin);
        auto lc = comp.lin_coeff(vin);
        int rhs = (uov::dot(oil, lc, q) + comp.vin_const(vin)) % q;
        CHECK(lhs == rhs, "eval_affine random failed");
    }
}

void test_eval_zero_oil() {
    const int q = 13, o = 3, v = 5;
    std::mt19937 rng(1);
    auto comp = uov::CentralMapComp::random(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec zero_oil(o, 0);
    for (int t = 0; t < 20; ++t) {
        uov::Vec vin(v);
        for (auto& x : vin) x = dist(rng);
        CHECK(comp.eval(zero_oil, vin) == comp.vin_const(vin), "eval_zero_oil failed");
    }
}

void test_eval_zero_vin() {
    const int q = 13, o = 3, v = 5;
    std::mt19937 rng(2);
    auto comp = uov::CentralMapComp::random(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec zero_vin(v, 0);
    for (int t = 0; t < 20; ++t) {
        uov::Vec oil(o);
        for (auto& x : oil) x = dist(rng);
        int expected = (uov::dot(comp.c, oil, q) + comp.e) % q;
        CHECK(comp.eval(oil, zero_vin) == expected, "eval_zero_vin failed");
    }
}

// ── eval_as_linSystem ─────────────────────────────────────────────────────────

void test_eval_as_lin_system() {
    const int q = 31, o = 4, v = 6;
    std::mt19937 rng(10);
    auto F = uov::CentralMap::random(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    for (int t = 0; t < 50; ++t) {
        uov::Vec oil(o), vin(v);
        for (auto& x : oil) x = dist(rng);
        for (auto& x : vin) x = dist(rng);
        auto lhs = F.eval(oil, vin);
        auto M = F.lin_matrix(vin);
        auto b = F.vin_const_vec(vin);
        auto Moil = uov::mat_mul_vec(M, oil, q);
        for (int k = 0; k < o; ++k)
            CHECK(lhs[k] == (Moil[k] + b[k]) % q, "eval_as_lin_system failed");
    }
}

// ── Gaussian elimination ──────────────────────────────────────────────────────

void test_gauss_identity() {
    const int q = 17, n = 4;
    uov::Mat I(n, uov::Vec(n, 0));
    for (int i = 0; i < n; ++i) I[i][i] = 1;
    uov::Vec b{3, 1, 4, 1};
    auto x = uov::gauss_solve(I, b, q);
    CHECK(x.has_value(), "gauss_solve failed on identity");
    CHECK(*x == b, "gauss_solve identity wrong result");
}

void test_gauss_random() {
    const int q = 31;
    std::mt19937 rng(42);
    std::uniform_int_distribution<int> dist(0, q - 1);
    for (int t = 0; t < 50; ++t) {
        int n = 2 + (rng() % 5);
        std::optional<uov::Mat> inv_opt;
        uov::Mat M;
        for (int attempt = 0; attempt < 20; ++attempt) {
            M = uov::random_mat(n, n, q, rng);
            inv_opt = uov::gf_matinv(M, q);
            if (inv_opt) break;
        }
        if (!inv_opt) continue;
        uov::Vec b(n);
        for (auto& x : b) x = dist(rng);
        auto x = uov::gauss_solve(M, b, q);
        CHECK(x.has_value(), "gauss_solve returned nullopt for invertible M");
        auto check = uov::mat_mul_vec(M, *x, q);
        CHECK(check == b, "gauss_solve solution wrong");
    }
}

void test_gauss_singular() {
    const int q = 7;
    uov::Mat M{{1, 2}, {2, 4}};  // rank 1
    auto x = uov::gauss_solve(M, {1, 2}, q);
    CHECK(!x.has_value(), "gauss_solve should return nullopt for singular M");
}

// ── UOV correctness ───────────────────────────────────────────────────────────

void test_roundtrip_many() {
    const int q = 31, o = 4, v = 8;
    for (int i = 0; i < 50; ++i) {
        std::mt19937 krng(i);
        auto key = uov::keygen(q, o, v, krng);
        std::uniform_int_distribution<int> dist(0, q - 1);
        std::mt19937 srng(i + 1000);
        uov::Vec msg(o);
        for (auto& x : msg) x = dist(srng);
        auto sig = key.sign(msg, srng);
        CHECK(sig.has_value(), "sign returned nullopt");
        CHECK(key.verify(msg, *sig), "verify failed after sign");
    }
}

void test_wrong_message_rejected() {
    const int q = 31, o = 4, v = 8;
    std::mt19937 rng(7);
    auto key = uov::keygen(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec msg(o);
    for (auto& x : msg) x = dist(rng);
    auto sig = key.sign(msg, rng);
    uov::Vec wrong = msg;
    for (auto& x : wrong) x = (x + 1) % q;
    CHECK(!key.verify(wrong, *sig), "wrong message should be rejected");
}

void test_forgery_rejected() {
    const int q = 31, o = 4, v = 8;
    std::mt19937 rng(77);
    auto key = uov::keygen(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec msg(o);
    for (auto& x : msg) x = dist(rng);
    int accepted = 0;
    for (int i = 0; i < 1000; ++i) {
        uov::Vec forgery(o + v);
        for (auto& x : forgery) x = dist(rng);
        if (key.verify(msg, forgery)) ++accepted;
    }
    CHECK(accepted == 0, "random forgery was accepted");
}

void test_cross_key_rejection() {
    const int q = 31, o = 4, v = 8;
    std::mt19937 rng1(10), rng2(11);
    auto key1 = uov::keygen(q, o, v, rng1);
    auto key2 = uov::keygen(q, o, v, rng2);
    std::mt19937 srng(55);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec msg(o);
    for (auto& x : msg) x = dist(srng);
    auto sig = key1.sign(msg, srng);
    CHECK(key1.verify(msg, *sig), "key1 verify failed");
    CHECK(!key2.verify(msg, *sig), "key2 should reject key1's signature");
}

void test_public_eval_consistent() {
    const int q = 31, o = 4, v = 8;
    std::mt19937 rng(1234);
    auto key = uov::keygen(q, o, v, rng);
    std::uniform_int_distribution<int> dist(0, q - 1);
    uov::Vec oil(o), vin(v);
    for (auto& x : oil) x = dist(rng);
    for (auto& x : vin) x = dist(rng);
    auto y = key.F.eval(oil, vin);
    uov::Vec combined(oil);
    combined.insert(combined.end(), vin.begin(), vin.end());
    auto sigma = uov::mat_mul_vec(key.T_inv, combined, q);
    CHECK(key.public_eval(sigma) == y, "public_eval inconsistent");
    CHECK(key.verify(y, sigma), "verify inconsistent with public_eval");
}

// ─────────────────────────────────────────────────────────────────────────────

int main() {
    std::cout << "UOV C++ correctness tests\n";

    RUN_TEST(test_eval_affine_exhaustive);
    RUN_TEST(test_eval_affine_random);
    RUN_TEST(test_eval_zero_oil);
    RUN_TEST(test_eval_zero_vin);
    RUN_TEST(test_eval_as_lin_system);
    RUN_TEST(test_gauss_identity);
    RUN_TEST(test_gauss_random);
    RUN_TEST(test_gauss_singular);
    RUN_TEST(test_roundtrip_many);
    RUN_TEST(test_wrong_message_rejected);
    RUN_TEST(test_forgery_rejected);
    RUN_TEST(test_cross_key_rejection);
    RUN_TEST(test_public_eval_consistent);

    std::cout << "\n" << passed << " passed, " << failed << " failed\n";
    return failed == 0 ? 0 : 1;
}

File: impl/python/tests/test_correctness.py
"""pytest suite — mirrors every Lean theorem in CentralMap.lean and SchemeCorrectness.lean."""

import random
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from uov import keygen
from uov.field import dot, mat_mul_vec, gauss_solve, gf_matinv
from uov.central_map import CentralMapComp, CentralMap


# ── helpers ──────────────────────────────────────────────────────────────────


def make_comp(q, o, v, seed):
    rng = random.Random(seed)
    return CentralMapComp.random(q, o, v, rng)


def make_key(q, o, v, seed=0):
    return keygen(q, o, v, seed=seed)


# ── eval_affine (CentralMapComp.eval_affine in Lean) ─────────────────────────


class TestEvalAffine:
    """eval(oil, vin) == dot(oil, linCoeff(vin)) + vinConst(vin) for all inputs."""

    def test_exhaustive_small(self):
        """Exhaustive check over GF(7)^2 × GF(7)^2 (2401 cases)."""
        q, o, v = 7, 2, 2
        comp = make_comp(q, o, v, seed=0)
        for oil0 in range(q):
            for oil1 in range(q):
                for vin0 in range(q):
                    for vin1 in range(q):
                        oil = [oil0, oil1]
                        vin = [vin0, vin1]
                        lhs = comp.eval(oil, vin)
                        rhs = (
                            dot(oil, comp.lin_coeff(vin), q) + comp.vin_const(vin)
                        ) % q
                        assert lhs == rhs, f"eval_affine failed at oil={oil} vin={vin}"

    def test_random_large(self):
        """100 random checks over GF(31) with o=4, v=8."""
        q, o, v = 31, 4, 8
        rng = random.Random(7)
        comp = make_comp(q, o, v, seed=3)
        for _ in range(100):
            oil = [rng.randrange(q) for _ in range(o)]
            vin = [rng.randrange(q) for _ in range(v)]
            lhs = comp.eval(oil, vin)
            rhs = (dot(oil, comp.lin_coeff(vin), q) + comp.vin_const(vin)) % q
            assert lhs == rhs

    def test_zero_oil(self):
        """At oil=0, eval equals vinConst."""
        q, o, v = 13, 3, 5
        comp = make_comp(q, o, v, seed=1)
        rng = random.Random(2)
        for _ in range(20):
            vin = [rng.randrange(q) for _ in range(v)]
            assert comp.eval([0] * o, vin) == comp.vin_const(vin)

    def test_zero_vin(self):
        """At vin=0, eval equals dot(c, oil) + e."""
        q, o, v = 13, 3, 5
        comp = make_comp(q, o, v, seed=2)
        rng = random.Random(3)
        for _ in range(20):
            oil = [rng.randrange(q) for _ in range(o)]
            expected = (dot(comp.c, oil, q) + comp.e) % q
            assert comp.eval(oil, [0] * v) == expected


# ── eval_as_linSystem (CentralMap.eval_as_linSystem in Lean) ─────────────────


class TestEvalAsLinSystem:
    """F(oil, vin) == M(vin)·oil + b(vin) for the full central map."""

    def test_random(self):
        q, o, v = 31, 4, 6
        rng = random.Random(10)
        F = CentralMap.random(q, o, v, rng)
        for _ in range(50):
            oil = [rng.randrange(q) for _ in range(o)]
            vin = [rng.randrange(q) for _ in range(v)]
            lhs = F.eval(oil, vin)
            M = F.lin_matrix(vin)
            b = F.vin_const_vec(vin)
            rhs = [(mat_mul_vec(M, oil, q)[k] + b[k]) % q for k in range(o)]
            assert lhs == rhs


# ── Gaussian elimination ──────────────────────────────────────────────────────


class TestGaussSolve:
    def test_identity(self):
        q = 17
        n = 4
        identity = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
        b = [3, 1, 4, 1]
        assert gauss_solve(identity, b, q) == b

    def test_random_systems(self):
        q = 31
        rng = random.Random(42)
        for _ in range(50):
            n = rng.randint(2, 6)
            inv_attempt = None
            for _ in range(20):
                M = [[rng.randrange(q) for _ in range(n)] for _ in range(n)]
                inv_attempt = gf_matinv(M, q)
                if inv_attempt is not None:
                    break
            if inv_attempt is None:
                continue
            b = [rng.randrange(q) for _ in range(n)]
            x = gauss_solve(M, b, q)
            assert x is not None
            check = mat_mul_vec(M, x, q)
            assert check == b

    def test_singular_returns_none(self):
        q = 7
        M = [[1, 2], [2, 4]]  # rank 1
        assert gauss_solve(M, [1, 2], q) is None


# ── UOV correctness theorem ───────────────────────────────────────────────────


class TestUOVCorrectness:
    """sign(sk, y) always passes verify(pk, y, ·)."""

    def test_roundtrip_many(self):
        """100 random key/message round trips over GF(31), o=4, v=8."""
        q, o, v = 31, 4, 8
        rng = random.Random(99)
        for i in range(100):
            key = make_key(q, o, v, seed=i)
            msg = [rng.randrange(q) for _ in range(o)]
            sig = key.sign(msg, rng)
            assert sig is not None, f"signing failed on iteration {i}"
            assert key.verify(msg, sig), f"verify failed on iteration {i}"

    def test_wrong_message_rejected(self):
        q, o, v = 31, 4, 8
        rng = random.Random(7)
        key = make_key(q, o, v, seed=5)
        msg = [rng.randrange(q) for _ in range(o)]
        sig = key.sign(msg, rng)
        wrong = [(x + 1) % q for x in msg]
        assert not key.verify(wrong, sig)

    def test_zero_vector_rejected(self):
        q, o, v = 31, 4, 8
        rng = random.Random(8)
        key = make_key(q, o, v, seed=6)
        msg = [rng.randrange(q) for _ in range(o)]
        assert not key.verify(msg, [0] * (o + v))

    def test_forgery_1000(self):
        """1000 random forgeries should all be rejected (astronomically likely)."""
        q, o, v = 31, 4, 8
        rng = random.Random(77)
        key = make_key(q, o, v, seed=7)
        msg = [rng.randrange(q) for _ in range(o)]
        n = o + v
        accepted = sum(
            key.verify(msg, [rng.randrange(q) for _ in range(n)]) for _ in range(1000)
        )
        assert accepted == 0

    def test_different_keys_same_message(self):
        """Signature from key1 must not verify under key2."""
        q, o, v = 31, 4, 8
        rng = random.Random(55)
        key1 = make_key(q, o, v, seed=10)
        key2 = make_key(q, o, v, seed=11)
        msg = [rng.randrange(q) for _ in range(o)]
        sig = key1.sign(msg, rng)
        assert key1.verify(msg, sig)
        assert not key2.verify(msg, sig)

    def test_small_field(self):
        """GF(2) — edge case where almost everything is 0 or 1."""
        q, o, v = 2, 2, 4
        rng = random.Random(0)
        key = make_key(q, o, v, seed=0)
        for msg in [[a, b] for a in range(2) for b in range(2)]:
            sig = key.sign(msg, rng)
            if sig is not None:
                assert key.verify(msg, sig)

    def test_parameters_gf7_o3_v5(self):
        q, o, v = 7, 3, 5
        rng = random.Random(33)
        key = make_key(q, o, v, seed=33)
        for _ in range(30):
            msg = [rng.randrange(q) for _ in range(o)]
            sig = key.sign(msg, rng)
            assert sig is not None
            assert key.verify(msg, sig)

    def test_public_eval_consistent(self):
        """public_eval(sign(oil, vin)) == F.eval(oil, vin) for known oil/vin."""
        q, o, v = 31, 4, 8
        rng = random.Random(1234)
        key = make_key(q, o, v, seed=22)
        oil = [rng.randrange(q) for _ in range(o)]
        vin = [rng.randrange(q) for _ in range(v)]
        y = key.F.eval(oil, vin)

        from uov.field import mat_mul_vec

        combined = oil + vin
        sigma = mat_mul_vec(key.T_inv, combined, q)

        assert key.public_eval(sigma) == y
        assert key.verify(y, sigma)
